src/container.ts

```diff
@@ -26,54 +29,114 @@ export type ContainerEventArgs = EventArgs | WindowEventArgs | LayoutEventArgs;
      * @param {MenuItem[]} menuItems (Optional) Context menu.
      */
-    addTrayIcon(details: TrayIconDetails, listener?: () => void, menuItems?: MenuItem[]);
+    public abstract addTrayIcon(details: TrayIconDetails, listener?: () => void, menuItems?: MenuItem[]);
+
+    public abstract showNotification(title: string, options?: NotificationOptions);
+
+    public abstract getAllWindows(): Promise<ContainerWindow[]>;
+
+    public static get ipc(): MessageBus {
+        return Container._ipc;
+    }
+
+    public static set ipc(value: MessageBus) {
+        Container._ipc = value;
+
+        // If a MessageBus exists for the container then redefine the static emitter to be based on that bus
+        // When a MessageBus is never defined, then the events only publish within current window
+        Container.staticEmitter = new EventEmitter(value);
+    }
 
     /**
      * A messaging bus for sending and receiving messages
      */
-    readonly ipc: MessageBus;
+    public get ipc(): MessageBus {
+        return Container.ipc;
+    }
+
+    public set ipc(value: MessageBus) {
+        Container.ipc = value;
+    }
 
     /**
      * Persistent storage
      */
-    storage: Storage;
+    public storage: Storage;
+
+    public addListener(eventName: ContainerEventType, listener: (event: ContainerEventArgs) => void): this { // tslint:disable-line
+        return super.addListener(eventName, listener);
+    }
+
+    public removeListener(eventName: ContainerEventType, listener: (event: ContainerEventArgs) => void): this { // tslint:disable-line
+        return super.removeListener(eventName, listener);
+    }
+
+    public emit(eventName: ContainerEventType, eventArgs: ContainerEventArgs) { // tslint:disable-line
+        super.emit(eventName, eventArgs);
+    }
+
+    public static addListener(eventName: ContainerEventType, listener: (event: ContainerEventArgs) => void) { // tslint:disable-line
+        if (Container.staticEmitter) { Container.staticEmitter.addListener(eventName, listener); }
+    }
+
+    public static removeListener(eventName: ContainerEventType, listener: (event: ContainerEventArgs) => void) { // tslint:disable-line
+        if (Container.staticEmitter) { Container.staticEmitter.removeListener(eventName, listener); }
+    }
+
+    public static emit(eventName: ContainerEventType, eventArgs: ContainerEventArgs) { // tslint:disable-line
+        if (Container.staticEmitter) { Container.staticEmitter.emit(eventName, eventArgs, Container.ipc); }
+    }
+
+    public static listeners(eventName: string): ((event: EventArgs) => void)[] { // tslint:disable-line
+        return Container.staticEmitter.listeners(eventName);
+    }
 }
 
 /**
  * Represents a common Container to be used as a base for any custom Container implementation.
  * @augments Container
  */
-export abstract class ContainerBase extends EventEmitter implements Container {
+export abstract class ContainerBase extends Container {
```

Comment:
All the abstract removals in here were because they were pulled up into Container.